compiler/rustc_serialize/src/opaque.rs

```diff
@@ -656,8 +663,12 @@ impl<'a> serialize::Decoder for Decoder<'a> {
     #[inline]
     fn read_str(&mut self) -> Result<Cow<'_, str>, Self::Error> {
         let len = self.read_usize()?;
-        let s = std::str::from_utf8(&self.data[self.position..self.position + len]).unwrap();
-        self.position += len;
+        let sentinel = self.data[self.position + len];
+        assert_eq!(sentinel, STR_SENTINEL);
```

Comment:
I found that `assert_eq!` seems to cause worse code generation than a simple `assert!`. Usually that's not a problem but in performance sensitive code it could make a difference. And in this particular case the better error message from `assert_eq!` doesn't really give additional information. So could you try just putting an `assert!` here?

Reply:
Done